Resolve directories on WorkQueue in UI process

szewai opened the pull request:
#### 3d45c70fab6dbc2f2f3810af811b32bdbc19be05
<pre>
Resolve directories on WorkQueue in UI process
<a href="https://bugs.webkit.org/show_bug.cgi?id=270300">https://bugs.webkit.org/show_bug.cgi?id=270300</a>
<a href="https://rdar.apple.com/123841003">rdar://123841003</a>

Reviewed by Chris Dumez.

Move resolving directoreis off main thread in UI process to avoid slowing down main thread. Main thread still needs to
wait directories to be resolved before using them, but now we dipsatch the task when WebsiteDataStore is created, so
there is chance that the directories are already resolved before the use. Also, the queue has the same priority as main
thread, so in the worst case where client uses WebsiteDataStore immediately after creating it, the performance will be
the same as current implementation.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _initWithConfiguration:]):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::gpuProcessSessionParameters):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
(WebKit::WebsiteDataStore::saveRecentSearches):
(WebKit::WebsiteDataStore::loadRecentSearches):
(WebKit::WebsiteDataStore::removeRecentSearches):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::createNonPersistent):
(WebKit::WebsiteDataStore::create):
(WebKit::WebsiteDataStore::WebsiteDataStore):
(WebKit::WebsiteDataStore::defaultDataStore):
(WebKit::migrateMediaKeysStorageIfNecessary):
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::resolvedDirectories const):
(WebKit::WebsiteDataStore::mediaKeysStorageSalt const):
(WebKit::WebsiteDataStore::resolveDirectoriesAsynchronously):
(WebKit::WebsiteDataStore::handleResolvedDirectoriesAsynchronously):
(WebKit::WebsiteDataStore::waitForDirectoriesToResolveIfNecessary):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::migrateMediaKeysStorageIfNecessary): Deleted.
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::perOriginStorageQuota const):
(WebKit::WebsiteDataStore::originQuotaRatio):
(WebKit::WebsiteDataStore::resolvedCacheStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedLocalStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedNetworkCacheDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedAlternativeServicesStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedMediaCacheDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedMediaKeysDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedJavaScriptConfigurationDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedCookieStorageFile const): Deleted.
(WebKit::WebsiteDataStore::resolvedIndexedDBDatabaseDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedServiceWorkerRegistrationDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedResourceLoadStatisticsDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedHSTSStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedGeneralStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedSearchFieldHistoryDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedModelElementCacheDirectory const): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::applicationCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setApplicationCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::mediaCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setMediaCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::mediaKeysStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setMediaKeysStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::alternativeServicesDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setAlternativeServicesDirectory):
(WebKit::WebsiteDataStoreConfiguration::javaScriptConfigurationDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setJavaScriptConfigurationDirectory):
(WebKit::WebsiteDataStoreConfiguration::searchFieldHistoryDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setSearchFieldHistoryDirectory):
(WebKit::WebsiteDataStoreConfiguration::indexedDBDatabaseDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setIndexedDBDatabaseDirectory):
(WebKit::WebsiteDataStoreConfiguration::webSQLDatabaseDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setWebSQLDatabaseDirectory):
(WebKit::WebsiteDataStoreConfiguration::hstsStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setHSTSStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::localStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setLocalStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::modelElementCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setModelElementCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::deviceIdHashSaltsStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setDeviceIdHashSaltsStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::cookieStorageFile const):
(WebKit::WebsiteDataStoreConfiguration::setCookieStorageFile):
(WebKit::WebsiteDataStoreConfiguration::resourceLoadStatisticsDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setResourceLoadStatisticsDirectory):
(WebKit::WebsiteDataStoreConfiguration::networkCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setNetworkCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::cacheStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setCacheStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::generalStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setGeneralStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::applicationCacheFlatFileSubdirectoryName const):
(WebKit::WebsiteDataStoreConfiguration::setApplicationCacheFlatFileSubdirectoryName):
(WebKit::WebsiteDataStoreConfiguration::serviceWorkerRegistrationDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setServiceWorkerRegistrationDirectory):
(WebKit::WebsiteDataStoreConfiguration::directories const):
* Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
* Source/WebKit/UIProcess/glib/WebProcessProxyGLib.cpp:
(WebKit::WebProcessProxy::platformGetLaunchOptions):

Canonical link: <a href="https://commits.webkit.org/275959@main">https://commits.webkit.org/275959@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab3603eddafa97df06299f230e4fd12a3502150c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43373 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39499 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19821 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35850 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43947 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37355 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16829 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1429 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38732 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47552 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18291 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42634 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19824 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41294 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9652 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20003 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19455 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -386,64 +392,115 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+    }
+
+    while (m_resolveState != ResolveState::Resolved)
+        m_resolveCondition.wait(m_resolveLock);
```
Hanging the high priority main thread on a low priority background thread doesn't seem like a good idea to me. Was this perf tested? This adds a lot of complexity and it isn't a clear win (since there is a code to dispatching back and forth between threads, and we'll end up hanging the main thread on the filesystem operations anyway if need the resolved directories very shortly after creating the data store (which likely we will). Not to mention the thread priority issue (but that's easily fixable.

szewai replied:
> Was this perf tested?

Looks like no performance impact on PLT; see link in the radar.

> we'll end up hanging the main thread on the filesystem operations anyway if need the resolved directories very shortly after creating the data store (which likely we will)

It depends on the use case. In the case where client initializes the data store but not use it for loading yet (e.g. creating a global data store when browser launches, like prewarming web process), this will be a perf improvement since the directories will be ready before using the data store, also main thread could keep responsive for user input. I noticed the file operations are slow when I investigated rdar://114388755 as they showed up in the profile.

As for the thread priority issue, I didn't notice much of the difference, probably because the system is not busy enough. But if you think it's better to bump the priority of the work queue to ensure main thread is unblocked, I could make the change.

cdumez replied:
> As for the thread priority issue, I didn't notice much of the difference, probably because the system is not busy enough. But if you think it's better to bump the priority of the work queue to ensure main thread is unblocked, I could make the change.

I don't think the PR should be merged without this. It is not OK to block the main thread on a lower priority thread. It partly defeats the purpose of this optimization.

Note that I don't know if ALL WKWebsiteDataStore background operations should be done on a high priority thread. However, it seems clear the directory resolving does, given that the main thread may block on the operation being done. I didn't check what other operations we do on that work queue but it's possible we'll need 2 work queues for e.g.

szewai replied:
Two work queues with different priorities seems fine; I will update the patch

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -142,16 +142,18 @@ Ref<WebsiteDataStore> WebsiteDataStore::createNonPersistent()
-    , m_trackingPreventionDebugMode(m_resolvedConfiguration->resourceLoadStatisticsDebugModeEnabled())
+    , m_trackingPreventionDebugMode(m_configuration->resourceLoadStatisticsDebugModeEnabled())
     , m_queue(WorkQueue::create("com.apple.WebKit.WebsiteDataStore"))
+    , m_resolveDirectoriesQueue(WorkQueue::create("com.apple.WebKit.WebsiteDataStore.resolveDirectories", WorkQueue::QOS::UserInteractive))
```
Do we really need a radar member? Presumably, we only need the queue once, shortly after construction and then the queue can go away. As a result, I'd recommend creating the queue in dispatchResolveDirectories() and dispatch on it there. The underlying dispatch queue will stay alive long enough to run its pending tasks without you needing to keep a a Ref to the WorkQueue. So the task WILL run.

szewai replied:
Sure, updated the patch with temporary queue in dispatchResolveDirectories.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +399,120 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+
+void WebsiteDataStore::dispatchResolveDirectories()
+{
+    if (!m_resolveDirectoriesQueue)
```
How can this happen?

szewai replied:
Removed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +399,120 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        return;
+
+    m_resolveDirectoriesQueue->dispatch([this, protectedThis = Ref { *this }, directories = m_configuration->directories().isolatedCopy()]() mutable {
+        auto resolvedDirectories = resolveDirectories(directories);
```
Maybe resolveDirectories() could resolve `directories` in place (i.e. modify its parameters rather than constructing a whole new structure). e.g.:
```
resolveDirectories(directories);
```

We have a local copy of the directories anyway.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +399,120 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        m_resolveState = ResolveState::Resolved;
+        m_resolveCondition.notifyOne();
+        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis)] { 
+            protectedThis->m_resolveDirectoriesQueue = nullptr;
```
We shouldn't need this with my proposal, though you'll likely still want to dispatch `protectedThis` back to the main thread.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +399,120 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        m_resolvedConfiguration->setDirectories(resolvedDirectories.isolatedCopy());
+        m_resolveState = ResolveState::Resolved;
+        m_resolveCondition.notifyOne();
+        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis)] { 
```
Probably this shouldn't be in the scope of the lock. Let's try and make the lock scope as tight as possible.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +399,120 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        return;
+
+    if (m_resolveState == ResolveState::Unresolved) {
+        dispatchResolveDirectories();
```
You dispatch in the create() function already, when is this useful? I see if don't dispatch for ephemeral stores but presumably that's because there is nothing to resolve?

szewai replied:
I removed it and replaced it with assertion.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +398,110 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+{
+    Locker resolveLocker { m_resolveLock };
+    m_resolveState = ResolveState::Resolving;
+    auto resolveDirectoriesQueue = WorkQueue::create("com.apple.WebKit.WebsiteDataStore.resolveDirectories", WorkQueue::QOS::UserInteractive);
```
Please use `Ref` instead of `auto` in the new. We try not to hide smart pointers anymore.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +398,110 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        resolveDirectories(directories);
+        {
+            Locker resolveLocker { m_resolveLock };
+            m_resolvedConfiguration->setDirectories(directories.isolatedCopy());
```
You want `crossThreadCopy(WTFMove(directories))` here. We have optimizations when cross-thread-copying certain types when we get a r-value reference. String in particular are optimized.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +398,110 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+
+void WebsiteDataStore::dispatchResolveDirectories()
+{
+    Locker resolveLocker { m_resolveLock };
```
This lock likely doesn't need to cover the WorkQueue creation or the dispatch?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -392,64 +398,110 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+    if (m_resolveState == ResolveState::Resolved)
+        return;
+
+    ASSERT(m_resolveState == ResolveState::Resolving);
```
We can probably `RELEASE_ASSERT()` here. I'd rather have a clear crash in release builds rather than a hang on the next line.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -545,12 +529,17 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
+    Condition m_resolveCondition;
+    enum class ResolveState : uint8_t { Unresolved, Resolving, Resolved };
+    ResolveState m_resolveState WTF_GUARDED_BY_LOCK(m_resolveLock) { ResolveState::Unresolved };
+    RefPtr<WebsiteDataStoreConfiguration> m_resolvedConfiguration WTF_GUARDED_BY_LOCK(m_resolveLock);
```
It is not clear to me which we still need a `m_resolvedConfiguration` data member with your change. Why don't we just have a `Directories m_resolvedDirectories` data member? It would also look better in terms of thread safety. Right now you have a lock on the `m_resolvedConfiguration` data member. But if someone somewhere outside WebsiteDataStore uses/modifies it, access won't be protected by the lock. Having a local struct of Strings instead of a RefPtr is much clearer for thread safety.

cdumez replied:
Maybe it can even be a `std::optional<Directories>` and then we can get rid of `ResolveState`? I don't really think we truly need 3 states?

cdumez reviewed (commented):
Will review some more later. Apologies for the review in chunks, this is a big patch.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
```diff
@@ -247,6 +247,33 @@ class WebsiteDataStoreConfiguration : public API::ObjectImpl<API::Object::Type::
+#if ENABLE(ARKIT_INLINE_PREVIEW)
+        String modelElementCacheDirectory;
+#endif
+        Directories isolatedCopy() const&;
```
Given the optimization I suggested above, and given that this struct only contains Strings (which can avoid copying when passing cross thread in some cases when we have a r-value reference), could you please add an `&&` overload (like we have for String).

cdumez reviewed (commented):
Ok, I think I did a complete review.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -580,6 +580,7 @@ - (instancetype)_initWithConfiguration:(_WKWebsiteDataStoreConfiguration *)confi
 
     auto sessionID = configuration.isPersistent ? PAL::SessionID::generatePersistentSessionID() : PAL::SessionID::generateEphemeralSessionID();
     API::Object::constructInWrapper<WebKit::WebsiteDataStore>(self, configuration->_configuration->copy(), sessionID);
+    _websiteDataStore->dispatchResolveDirectories();
```
This should be in the constructor / factory function, like it used to be. Too likely to forget to call otherwise.

szewai replied:
We cannot create Ref for WebsiteDataStore in constructor because it's not RefCounted. It inherits APIObject which calls CFRetain for ref(). It would crash inside CFRetain with:
```
Thread 0 Crashed:
0   CoreFoundation                	       0x180e3df50 CFRetain.cold.1 + 16
1   CoreFoundation                	       0x180c5a6e0 CFRetain + 140
2   WebKit                        	       0x131a2814c API::Object::ref() const + 32 (APIObject.mm:126)
3   WebKit                        	       0x132a2fb9c WTF::DefaultRefDerefTraits<WebKit::WebsiteDataStore>::ref(WebKit::WebsiteDataStore&) + 28 (Ref.h:55)
4   WebKit                        	       0x132a2fb68 WTF::Ref<WebKit::WebsiteDataStore, WTF::RawPtrTraits<WebKit::WebsiteDataStore>, WTF::DefaultRefDerefTraits<WebKit::WebsiteDataStore>>::Ref(WebKit::WebsiteDataStore&) + 40 (Ref.h:86)
5   WebKit                        	       0x132972970 WTF::Ref<WebKit::WebsiteDataStore, WTF::RawPtrTraits<WebKit::WebsiteDataStore>, WTF::DefaultRefDerefTraits<WebKit::WebsiteDataStore>>::Ref(WebKit::WebsiteDataStore&) + 40 (Ref.h:87)
6   WebKit                        	       0x132e5f308 WebKit::WebsiteDataStore::dispatchResolveDirectories() + 200 (WebsiteDataStore.cpp:472)
7   WebKit                        	       0x132e5ede8 WebKit::WebsiteDataStore::WebsiteDataStore(WTF::Ref<WebKit::WebsiteDataStoreConfiguration, WTF::RawPtrTraits<WebKit::WebsiteDataStoreConfiguration>, WTF::DefaultRefDerefTraits<WebKit::WebsiteDataStoreConfiguration>>&&, PAL::SessionID) + 712 (WebsiteDataStore.cpp:162)
8   WebKit                        	       0x132e5eab8 WebKit::WebsiteDataStore::WebsiteDataStore(WTF::Ref<WebKit::WebsiteDataStoreConfiguration, WTF::RawPtrTraits<WebKit::WebsiteDataStoreConfiguration>, WTF::DefaultRefDerefTraits<WebKit::WebsiteDataStoreConfiguration>>&&, PAL::SessionID) + 48 (WebsiteDataStore.cpp:159)
9   WebKit                        	       0x131ed0f88 void API::Object::constructInWrapper<WebKit::WebsiteDataStore, WTF::Ref<WebKit::WebsiteDataStoreConfiguration, WTF::RawPtrTraits<WebKit::WebsiteDataStoreConfiguration>, WTF::DefaultRefDerefTraits<WebKit::WebsiteDataStoreConfiguration>>, PAL::SessionID&>(id<WKObject>, WTF::Ref<WebKit::WebsiteDataStoreConfiguration, WTF::RawPtrTraits<WebKit::WebsiteDataStoreConfiguration>, WTF::DefaultRefDerefTraits<WebKit::WebsiteDataStoreConfiguration>>&&, PAL::SessionID&) + 84 (APIObject.h:237)
```

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -236,12 +239,11 @@ Ref<WebsiteDataStore> WebsiteDataStore::defaultDataStore()
-        isPersistent == IsPersistent::Yes ? PAL::SessionID::defaultSessionID() : PAL::SessionID::generateEphemeralSessionID()));
-    globalDatasStore = newDataStore.get();
-    protectedDefaultDataStore() = newDataStore.get();
+    auto newDataStore = WebsiteDataStore::create(WebsiteDataStoreConfiguration::create(isPersistent), isPersistent == IsPersistent::Yes ? PAL::SessionID::defaultSessionID() : PAL::SessionID::generateEphemeralSessionID());
```
`auto` -> `Ref`

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,117 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+WebsiteDataStoreConfiguration::Directories WebsiteDataStore::resolvedDirectories()
+{
+    Locker resolveLocker { m_resolveLock };
+    auto result = m_resolvedDirectories;
```
I think we can just `return *m_resolvedDirectories;`. The assertion isn't super valuable because std::optional also asserts the same thing.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,117 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        resolveDirectories(directories);
+        {
+            Locker resolveLocker { m_resolveLock };
+            m_resolvedDirectories = crossThreadCopy(directories);
```
`crossThreadCopy(WTFMove(directories))`

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,117 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        {
+            Locker resolveLocker { m_resolveLock };
+            m_resolvedDirectories = crossThreadCopy(directories);
+            m_resolveCondition.notifyOne();
```
`m_resolveCondition` is too generic a name. maybe `m_resolveDirectoriesCondition`?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
```diff
@@ -247,6 +247,33 @@ class WebsiteDataStoreConfiguration : public API::ObjectImpl<API::Object::Type::
+        Directories isolatedCopy() const&;
+        Directories isolatedCopy() &&;
+    };
+    Directories directories() const { return m_directories; }
```
Should return a `const Directories&`

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,117 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
-        Vector<String> allCacheDirectories = {
-            resolvedMediaCacheDirectory()
-            , resolvedNetworkCacheDirectory()
+WebsiteDataStoreConfiguration::Directories WebsiteDataStore::resolvedDirectories()
```
Given that `m_resolvedDirectories` MUST have been set by the bg thread before this function has been called, as given that the bg thread can never overwrite the data member later on, I believe it is safe to return a `WebsiteDataStoreConfiguration::Directories`. Also, the member function should be marked as `const`.

Finally, every call site can store in an `auto&` instead of `auto`.

szewai replied:
I think the reason it cannot be marked as const is the lock; I will try making the change and check.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,117 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+    });
+}
+
+void WebsiteDataStore::didResolveDirectories()
```
Could you explain why this must be called on the main thread? This wasn't obvious to me.
Ideally, we'd call this straight from the resolve queue.

szewai replied:
Because we are dispatching the tasks to a work queue with lower priority. The resolve directories queue has the same priority as main thread to avoid blocking main thread, but these tasks don't need that.

szewai replied:
oh you are asking why we cannot do the task dispatch on the dispatch work queue, not why we do the task dispatch. Yes I think that's possible; will change.

cdumez replied:
Correct. You can dispatch to your lower priority queue straight from the high priority queue, without involving the main thread.

cdumez commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -644,15 +644,16 @@ static inline GPUProcessSessionParameters gpuProcessSessionParameters(const Webs
     GPUProcessSessionParameters parameters;
 
-    parameters.mediaCacheDirectory = store.resolvedMediaCacheDirectory();
+    auto resolvedDirectories = const_cast<WebsiteDataStore&>(store).resolvedDirectories();
```
Why are you using `auto&` here as I suggested in my earlier review?

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -820,7 +820,8 @@ WebProcessDataStoreParameters WebProcessPool::webProcessDataStoreParameters(WebP
     websiteDataStore.resolveDirectoriesIfNecessary();
 
-    String mediaCacheDirectory = websiteDataStore.resolvedMediaCacheDirectory();
+    auto resolvedDirectories = websiteDataStore.resolvedDirectories();
```
Why are you using `auto&` here?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -179,9 +179,10 @@ bool experimentalFeatureEnabled(const String& key, bool defaultValue)
     if (!httpsProxy.isValid() && (isSafari || isMiniBrowser))
         httpsProxy = URL { [defaults stringForKey:(NSString *)WebKit2HTTPSProxyDefaultsKey] };
 
+    auto directories = resolvedDirectories();
```
Why are you using `auto&` here?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,114 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
     }
+}
+
+const WebsiteDataStoreConfiguration::Directories& WebsiteDataStore::resolvedDirectories()
```
As mentioned in my earlier review, it would be good to mark this method as `const`. You can achieve this by marking the `Lock` as `mutable` like we often do.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -543,12 +528,16 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
     const PAL::SessionID m_sessionID;
 
-    Ref<WebsiteDataStoreConfiguration> m_resolvedConfiguration;
+    Lock m_resolveLock;
```
Please mark as `mutable`, also rename to `m_resolvedDirectoriesLock`.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -543,12 +528,16 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
 #endif
 
     String migrateMediaKeysStorageIfNecessary(const String& directory);
+    void resolveDirectories(WebsiteDataStoreConfiguration::Directories&);
```
Can this function be marked as `static`? I took a quick look at the implementation and it seems like maybe?
If so, could it at least be marked as `const`?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,114 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
-        Vector<String> allCacheDirectories = {
-            resolvedMediaCacheDirectory()
-            , resolvedNetworkCacheDirectory()
+    m_hasDispatchedResolveDirectories = true;
```
Should we `ASSERT(!m_hasDispatchedResolveDirectories);` before this?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,114 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+void WebsiteDataStore::didResolveDirectories()
+{
+    Vector<String> allCacheDirectories;
+    auto directories = resolvedDirectories();
```
Can we use `auto&` here? Note that you're off the main thread now so copying the struct and thus the Strings it holds is likely not safe.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,114 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+        // Default paths of WebsiteDataStore created with identifier are not under caches or tmp directory,
+        // so we need to explicitly exclude them from backup.
+        allCacheDirectories = {
+            directories.mediaCacheDirectory
```
Since you're not on the main thread, you likely need to call `isolatedCopy()` on each String in `directories` that you want to copy. Otherwise, you end up with ref/deref of Strings on the main thread and this thread. It isn't safe.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -395,82 +397,114 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
-    if (auto webSQLDirectory = configuration->webSQLDatabaseDirectory(); !webSQLDirectory.isEmpty()) {
-        protectedQueue()->dispatch([webSQLDirectory = webSQLDirectory.isolatedCopy()]() {
+    // Clear data of deprecated types.
+    protectedQueue()->dispatch([webSQLDirectory = crossThreadCopy(directories.webSQLDatabaseDirectory), applicationCacheDirectory = crossThreadCopy(directories.applicationCacheDirectory), applicationCacheFlatFileSubdirectoryName = crossThreadCopy(directories.applicationCacheFlatFileSubdirectoryName), directoriesToExclude = crossThreadCopy(WTFMove(allCacheDirectories))]() {
```
If you call `isolatedCopy()` on the Strings you're using to initialize `allCacheDirectories` above (as I suggested) then you could get rid of the `crossThreadCopy()` for `allCacheDirectories` here and just use `WTFMove()`.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1845,15 +1897,16 @@ WebsiteDataStoreParameters WebsiteDataStore::parameters()
     resolveDirectoriesIfNecessary();
 
-    auto resourceLoadStatisticsDirectory = resolvedResourceLoadStatisticsDirectory();
+    auto directories = resolvedDirectories();
```
Why not `auto&` to avoid copying so many Strings?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -484,6 +468,7 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
     void setRestrictedOpenerTypeForDomainForTesting(const WebCore::RegistrableDomain&, RestrictedOpenerType);
 
     bool operator==(const WebsiteDataStore& other) const { return (m_sessionID == other.sessionID()); }
+    void dispatchResolveDirectories();
```
May be nicer to call this `revolveDirectoriesAsynchronously()`

cdumez commented on Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp:
```diff
@@ -35,13 +35,14 @@ namespace WebKit {
 void WebsiteDataStore::platformSetNetworkParameters(WebsiteDataStoreParameters& parameters)
 {
-    auto alternativeServiceStorageDirectory = resolvedAlternativeServicesStorageDirectory();
+    auto directories = resolvedDirectories();
```
`auto&` ?

cdumez reviewed (commented):
No r+ yet because I believe there are some thread-safety bugs.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -818,9 +818,10 @@ void WebProcessPool::registerHighDynamicRangeChangeCallback()
 WebProcessDataStoreParameters WebProcessPool::webProcessDataStoreParameters(WebProcessProxy& process, WebsiteDataStore& websiteDataStore)
 {
-    websiteDataStore.resolveDirectoriesIfNecessary();
+    websiteDataStore.waitDirectoriesToResolveIfNecessary();
```
I would rename this to `waitForDirectoriesToResolveIfNecessary()`. It sounds a little weird otherwise.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -137,20 +137,23 @@ void WebsiteDataStore::forEachWebsiteDataStore(Function<void(WebsiteDataStore&)>
 Ref<WebsiteDataStore> WebsiteDataStore::createNonPersistent()
 {
-    return adoptRef(*new WebsiteDataStore(WebsiteDataStoreConfiguration::create(IsPersistent::No), PAL::SessionID::generateEphemeralSessionID()));
+    auto result = adoptRef(*new WebsiteDataStore(WebsiteDataStoreConfiguration::create(IsPersistent::No), PAL::SessionID::generateEphemeralSessionID()));
```
Please use `Ref` instead of `auto`. We prefer not to hide smart pointers anymore.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -137,20 +137,23 @@ void WebsiteDataStore::forEachWebsiteDataStore(Function<void(WebsiteDataStore&)>
 Ref<WebsiteDataStore> WebsiteDataStore::create(Ref<WebsiteDataStoreConfiguration>&& configuration, PAL::SessionID sessionID)
 {
-    return adoptRef(*new WebsiteDataStore(WTFMove(configuration), sessionID));
+    auto result = adoptRef(*new WebsiteDataStore(WTFMove(configuration), sessionID));
```
`Ref`

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -386,91 +388,135 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+            Locker resolveLocker { m_resolveDirectoriesLock };
+            m_resolvedDirectories = crossThreadCopy(WTFMove(directories));
+            m_resolveDirectoriesCondition.notifyOne();
+            m_mediaKeysStorageSalt = WTFMove(mediaKeysStorageSalt);
```
Are we sure it is OK to do this *after* notifying the condition?

szewai replied:
I think yes because lock is not released yet; will move it up to avoid confusion.

cdumez reviewed (approved):
LGTM with nits.

szewai committed:
Resolve directories on WorkQueue in UI process
https://bugs.webkit.org/show_bug.cgi?id=270300
rdar://123841003

Reviewed by Chris Dumez.

Move resolving directoreis off main thread in UI process to avoid slowing down main thread. Main thread still needs to
wait directories to be resolved before using them, but now we dipsatch the task when WebsiteDataStore is created, so
there is chance that the directories are already resolved before the use. Also, the queue has the same priority as main
thread, so in the worst case where client uses WebsiteDataStore immediately after creating it, the performance will be
the same as current implementation.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _initWithConfiguration:]):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::gpuProcessSessionParameters):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
(WebKit::WebsiteDataStore::saveRecentSearches):
(WebKit::WebsiteDataStore::loadRecentSearches):
(WebKit::WebsiteDataStore::removeRecentSearches):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::createNonPersistent):
(WebKit::WebsiteDataStore::create):
(WebKit::WebsiteDataStore::WebsiteDataStore):
(WebKit::WebsiteDataStore::defaultDataStore):
(WebKit::migrateMediaKeysStorageIfNecessary):
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::resolvedDirectories const):
(WebKit::WebsiteDataStore::mediaKeysStorageSalt const):
(WebKit::WebsiteDataStore::resolveDirectoriesAsynchronously):
(WebKit::WebsiteDataStore::handleResolvedDirectoriesAsynchronously):
(WebKit::WebsiteDataStore::waitForDirectoriesToResolveIfNecessary):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::migrateMediaKeysStorageIfNecessary): Deleted.
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::perOriginStorageQuota const):
(WebKit::WebsiteDataStore::originQuotaRatio):
(WebKit::WebsiteDataStore::resolvedCacheStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedLocalStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedNetworkCacheDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedAlternativeServicesStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedMediaCacheDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedMediaKeysDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedJavaScriptConfigurationDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedCookieStorageFile const): Deleted.
(WebKit::WebsiteDataStore::resolvedIndexedDBDatabaseDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedServiceWorkerRegistrationDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedResourceLoadStatisticsDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedHSTSStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedGeneralStorageDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedSearchFieldHistoryDirectory const): Deleted.
(WebKit::WebsiteDataStore::resolvedModelElementCacheDirectory const): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::applicationCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setApplicationCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::mediaCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setMediaCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::mediaKeysStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setMediaKeysStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::alternativeServicesDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setAlternativeServicesDirectory):
(WebKit::WebsiteDataStoreConfiguration::javaScriptConfigurationDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setJavaScriptConfigurationDirectory):
(WebKit::WebsiteDataStoreConfiguration::searchFieldHistoryDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setSearchFieldHistoryDirectory):
(WebKit::WebsiteDataStoreConfiguration::indexedDBDatabaseDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setIndexedDBDatabaseDirectory):
(WebKit::WebsiteDataStoreConfiguration::webSQLDatabaseDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setWebSQLDatabaseDirectory):
(WebKit::WebsiteDataStoreConfiguration::hstsStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setHSTSStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::localStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setLocalStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::modelElementCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setModelElementCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::deviceIdHashSaltsStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setDeviceIdHashSaltsStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::cookieStorageFile const):
(WebKit::WebsiteDataStoreConfiguration::setCookieStorageFile):
(WebKit::WebsiteDataStoreConfiguration::resourceLoadStatisticsDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setResourceLoadStatisticsDirectory):
(WebKit::WebsiteDataStoreConfiguration::networkCacheDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setNetworkCacheDirectory):
(WebKit::WebsiteDataStoreConfiguration::cacheStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setCacheStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::generalStorageDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setGeneralStorageDirectory):
(WebKit::WebsiteDataStoreConfiguration::applicationCacheFlatFileSubdirectoryName const):
(WebKit::WebsiteDataStoreConfiguration::setApplicationCacheFlatFileSubdirectoryName):
(WebKit::WebsiteDataStoreConfiguration::serviceWorkerRegistrationDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setServiceWorkerRegistrationDirectory):
(WebKit::WebsiteDataStoreConfiguration::directories const):
* Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
* Source/WebKit/UIProcess/glib/WebProcessProxyGLib.cpp:
(WebKit::WebProcessProxy::platformGetLaunchOptions):

Canonical link: https://commits.webkit.org/275959@main
